packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js

```diff
@@ -1381,50 +1399,97 @@ function collectRecommendations({
+  switch (node.type) {
+    case 'ObjectExpression':
+      return 'object';
+    case 'ArrayExpression':
+      return 'array';
+    case 'ArrowFunctionExpression':
+    case 'FunctionExpression':
+      return 'function';
+    case 'ClassExpression':
+      return 'class';
+    case 'ConditionalExpression':
+      if (
+        getConstructionExpresionType(node.consequent) != null ||
+        getConstructionExpresionType(node.alternate) != null
+      ) {
+        return 'conditional';
+      }
+      return null;
+    case 'LogicalExpression':
+      if (
+        getConstructionExpresionType(node.left) != null ||
+        getConstructionExpresionType(node.right) != null
+      ) {
+        return 'logical expression';
+      }
+      return null;
+  }
+  return null;
+}
+
+// Finds variables declared as dependencies
 // that would invalidate on every render.
-function scanForDeclaredBareFunctions({
+function scanForConstructions({
   declaredDependencies,
   declaredDependenciesNode,
   componentScope,
   scope,
 }) {
-  const bareFunctions = declaredDependencies
+  const constructions = declaredDependencies
     .map(({key}) => {
-      const fnRef = componentScope.set.get(key);
-      if (fnRef == null) {
+      const ref = componentScope.variables.find(v => v.name === key);
+      if (ref == null) {
+        return null;
+      }
+
+      if (ref.references.some(r => !r.init && r.isWrite())) {
+        // The variable gets reassigned. This complicates things so we won't
+        // try to reason about it for now.
         return null;
       }
-      const fnNode = fnRef.defs[0];
-      if (fnNode == null) {
+
+      const node = ref.defs[0];
+      if (node == null) {
         return null;
       }
       // const handleChange = function () {}
       // const handleChange = () => {}
+      // const foo = {}
+      // const foo = []
+      // etc.
       if (
-        fnNode.type === 'Variable' &&
-        fnNode.node.type === 'VariableDeclarator' &&
-        fnNode.node.init != null &&
-        (fnNode.node.init.type === 'ArrowFunctionExpression' ||
-          fnNode.node.init.type === 'FunctionExpression')
+        node.type === 'Variable' &&
+        node.node.type === 'VariableDeclarator' &&
+        node.node.id.type === 'Identifier' && // Ensure this is not destructed assignment
+        node.node.init != null
       ) {
-        return fnRef;
+        const constantExpressionType = getConstructionExpresionType(
+          node.node.init,
+        );
+        if (constantExpressionType != null) {
+          return [ref, constantExpressionType];
+        }
       }
       // function handleChange() {}
       if (
-        fnNode.type === 'FunctionName' &&
-        fnNode.node.type === 'FunctionDeclaration'
+        node.type === 'FunctionName' &&
+        node.node.type === 'FunctionDeclaration'
       ) {
-        return fnRef;
+        return [ref, 'function'];
       }
       return null;
```

Comment:
would be an exceedingly rare case, but could also add handling for class declarations:
```js
function Component() {
  class WhyTho {}
  useMemo(() => {}, [WhyTho]);
}
```